Adding multi-os and multi-version signatures support

aminya opened the pull request:
- [X] Redesigned the bot
- [x] using functions instead of macros for more stability
	- [x] get the path to the example_script
	- [x] automatic search for the script
        - [x] combined the snooping methods together
- [x] more stability for deactivation/activation
- Many new configurations and features such as:
- [X] Multiple OS signature generation support
- [x] Multiple Julia version support
- [X] YAML file: creating one consolidated PR for all of the OS/Versions
- [x] being resilient to the current directory
- [x] accepting precompile root path
- [x] accepting package path
- [x] add optional exhaustive checker (`try catch eval` all)
- [x] blacklist `Main` by default
- [x] add `SnoopCompile_ENV `for skipping a test in SnoopCompile environment
- [x] many error checks and warnings/info messages
- [x] tests (bot tests pass locally)
- [x] using GitHub actions for CI (except for docs)

Fixes https://github.com/timholy/SnoopCompile.jl/issues/65
Fixes https://github.com/timholy/SnoopCompile.jl/issues/55
Fixes https://github.com/timholy/SnoopCompile.jl/issues/62
Fixes #76 
Fixes https://github.com/timholy/SnoopCompile.jl/issues/84
Fixes https://github.com/timholy/SnoopCompile.jl/issues/80
Fixes #78 


# Review Tips
- I think running the tests is the easiest way for reviewing (test/bot.jl)
- This PR got huge over time. I can give a walkthrough in a Skype call if it is needed.

aminya commented:
@timholy I decided to have a `@static if` that we put all the precompile scripts inside. 
However, to keep things clean, I am wondering if we can write the code in a separate file and then include it. But does  `@static` work the same way inside that file as it was outside in the main file of a package?

The file can be something like `src/precompile_includer.jl`.

timholy commented:
> But does @static work the same way inside that file as it was outside in the main file of a package?

Should be the same no matter where you put it.

aminya commented:
I am running this on Example.jl in locally (https://github.com/aminya/Example.jl) and the macro generates this code:
```julia
# precompile_enclusre (don't edit the following!)
should_precompile =   true
ismultios = true
@static if should_precompile
    @static if !ismultios
        include("../deps/SnoopCompile/precompile/precompile_Example.jl")
    elseif Sys.iswindows
         include("../deps/SnoopCompile/precompile/windows/precompile_Example.jl")
elseif Sys.islinux
         include("../deps/SnoopCompile/precompile/linux/precompile_Example.jl")

    end
end # precompile_enclusure
_precompile_()
```

But I get this error when I run it:
```julia
LoadError: MethodError: no method matching xor(::typeof(Base.Sys.iswindows), ::Bool)
Closest candidates are:
  xor(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:529
  xor(!Matched::Bool, ::Bool) at bool.jl:75
  xor(!Matched::Missing, ::Bool) at missing.jl:165

```

aminya commented:
So as of now, the feature works! A PR is successfully generated on Linux:
https://github.com/aminya/Example.jl/actions/runs/40199492

However, the workflow file isn't parsed correctly on Windows.
https://github.com/aminya/Example.jl/runs/448862749?check_suite_focus=true#step:5:1

aminya commented:
> 
> # Windows Fails
> 
> However, the workflow file isn't parsed correctly on Windows.
> [aminya/Example.jl/runs/448862749?check_suite_focus=true#step:5:1](https://github.com/aminya/Example.jl/runs/448862749?check_suite_focus=true#step:5:1)
> 
> This is an issue with PowerShell [MicrosoftDocs/PowerShell-Docs#2361](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/2361)


The issue is resolved by using `shell:bash` 🚀

timholy commented on src/bot.jl:
```diff
@@ -8,24 +8,36 @@ const UStrings = Union{AbstractString,Regex,AbstractChar}
+
+- `blacklist` : A vector of of Strings (or RegExp) to remove some precompile sentences
+
+- `os`: A vector of of Strings (or RegExp) to give the list of os that you want to generate precompile signatures for. Each element will call a `Sys.is\$eachos()` function.
```
One disadvantage is that this doesn't clarify whether these are positional or keyword arguments. The usual way I might think of writing this kind of documentation would be

```julia
    BotConfig(package_name::AbstractString; subst = [], blacklist = [], os = [])

Construct a SnoopCompile bot configuration. `package_name` is the name of the package. You may supply one or more of the following keyword arguments:
...
```

timholy commented on src/bot.jl:
```diff
@@ -8,24 +8,36 @@ const UStrings = Union{AbstractString,Regex,AbstractChar}
 struct BotConfig
-    packageName::String
+    package_name::String
     subst::Vector{Pair{T1, T2}} where {T1<:UStrings, T2 <: UStrings}
```
```suggestion
    subst::Vector{Pair{UStrings, UStrings}}
```
The compiler won't get any benefit from your parametric form since the parameters don't propagate to `BotConfig` itself. By making it broader you would allow one to push new items onto the list without worrying about whether the original type was incompatibility (e.g., if you create it with a `String=>String` type and then push on a Regex-based pair).

timholy commented on src/bot.jl:
```diff
@@ -8,24 +8,36 @@ const UStrings = Union{AbstractString,Regex,AbstractChar}
-    packageName::String
+    package_name::String
     subst::Vector{Pair{T1, T2}} where {T1<:UStrings, T2 <: UStrings}
     blacklist::Vector{T3} where {T3<:UStrings}
```
```suggestion
    blacklist::Vector{UStrings}
```

timholy commented on src/bot.jl:
```diff
@@ -8,24 +8,36 @@ const UStrings = Union{AbstractString,Regex,AbstractChar}
 
-function BotConfig(packageName::String; subst::Vector{Pair{T1, T2}} where {T1<:UStrings, T2 <: UStrings} = Vector{Pair{String, String}}(), blacklist::Vector{T3} where {T3<:UStrings}= String[])
-    return BotConfig(packageName, subst, blacklist)
+function BotConfig(package_name::String; subst::Vector{Pair{T1, T2}} where {T1<:UStrings, T2 <: UStrings} = Vector{Pair{String, String}}(), blacklist::Vector{T3} where {T3<:UStrings}= String[], os::Union{Vector{String}, Nothing} = nothing)
```
```suggestion
function BotConfig(package_name::AbstractString; subst::AbstractVector = Vector{Pair{UString, UString}}(), blacklist::AbstractVector= UString[], os::Union{Vector{String}, Nothing} = nothing)
```
If the user passes in some other kind of compatible object, Julia will convert it for you.

timholy commented on src/bot/(de)activator.jl:
```diff
@@ -0,0 +1,48 @@
+
+Activates precompile of a package by setting `should_precompile = true`
+
+`package_path` is the same as `pathof`. However, `pathof(module)` isn't used to prevent loadnig the package.
```
```suggestion
`package_path` should be the full path to the defining file for the package, i.e., identical to `pathof(ThePkg)`. However, `pathof(module)` isn't used to prevent the need to load the package.
```

timholy commented on src/bot/(de)activator.jl:
```diff
@@ -0,0 +1,48 @@
+
+Deactivates precompile of a package by setting `should_precompile = false`
+
+`package_path` is the same as `pathof`. However, `pathof(module)` isn't used to prevent loadnig the package.
```
Same change as above.

timholy commented on src/bot/precompile_include.jl:
```diff
@@ -0,0 +1,144 @@
+
+Creates a "precompile_includer.jl" file.
+"""
+function new_includer_file(package_name::String, package_path::String, os::Union{Vector{String}, Nothing})
```
```suggestion
function new_includer_file(package_name::AbstractString, package_path::AbstractString, os::Union{Vector{String}, Nothing})
```

timholy commented on src/bot/precompile_include.jl:
```diff
@@ -0,0 +1,144 @@
+    end
+
+    @info "$includer_file file will be created/overwritten"
+    enclusure = """
```
The word "enclusure" seems a bit unusual, what does it mean? Google isn't helping much.

aminya replied:
Sorry! That's a typo. 😄

timholy commented on src/bot/precompile_include.jl:
```diff
@@ -0,0 +1,144 @@
+        ## find end of a module
+        # assumes that the last `end` is the end of a module
+        endline = length(package_lines)
+        for iLine = endline:1
```
```suggestion
        for iLine = endline:-1:1
```

aminya commented:
The things I changed screwed up something. I get an unexpectend `end`. I can't find the issue.

aminya commented:
The Appveyor error is very strange and it doesn't happen offline. I am going to Github actions to see if we get the same error there

aminya commented:
I added a couple of features:
- else_os
- multi-version
- else_version

This is the logic. It was very complicated 🤓 . I haven't tested it yet 😄

aminya commented:
ColorTypes tests fail which isn't related to this PR.

aminya commented:
This is ready for review. @timholy 

I will address the YAML file issue for only creating one PR later.

aminya commented:
See the running actions here: https://github.com/aminya/Example.jl/actions/runs/81722039

aminya commented:
ColorTypes tests fail (its actual package tests which are used here), which is not related to this PR.

I am not going to touch this anymore, so I don't change anything unless you ask me. I don't want to get tempted to add a new feature and make this bigger. 😄

timholy commented:
Sitting down to review this now (with a nice cup of tea to help :wink:). Let's start by closing & opening to retrigger the tests, since the ColorTypes problem on nightly has been fixed.

aminya commented:
I see that some of the tests fail in Julia 1.0. Let me check them.

aminya commented:
Ahh. That is the issue of printing `nothing` in older Julias. I made an issue for it in Compat. https://github.com/JuliaLang/Compat.jl/issues/698. I don't have a solution for it, but I can limit Julia bot to run over 1.2.
```
  ArgumentError: `nothing` should not be printed; use `show`, `repr`, or custom output instead.
  Stacktrace:
```

aminya commented:
OK. The tests are fixed. 😄

aminya commented:
> OK, here's a first pass anyway. This seems like a much cleaner approach than editing the developer's main toplevel file. Nice!
> 

Yes! The previous version was kind of hacky.

> One concern is, why so many non-recommended variants of `@snoopi_bench` and `@snoopi_bot`? Is this all about deprecation? I want to quickly move to having one clean way of doing things. I'm supportive of this PR because it's clearly moving in the right direction, but it also validates my unease about ever having merged this in the first place.
> 

I was using macros at first, but I soon realized they are less stable than functions. For example, if we use another macro in front of them, Julia tries to expand the inner macro before `using Zygote`!
```jl
@snoopi_bot BotConfig("Zygote") begin
	using Zygote
    @code_adjoint ((x, y) -> sum(x.*y))([1,2,3],[4,5,6]) 
end
```
But these issues don't exist when we include the file instead. To have the same functionality, I made a macro-like function that accepts the simple scripts as Expr.

If you want, we can deprecate the macros right now in the coming version.

> We should consider putting the bot utils in a separate package if they aren't specific to SnoopCompile. `timesum` is, obviously, but the rest seem generic.
> 
Yeah, I kind of agree. 
- `addtestdep` should go to Pkg
- `GoodPath`, search file stuff can go to `FilePathsBase`
- `detectOS` and `standardize_osname` could be standard Julia functions
- `pathof_noload` could be a standard Julia function too

aminya commented on src/bot/precompile_include.jl:
```diff
@@ -0,0 +1,247 @@
+        os::Union{Vector{String}, Nothing},
+        else_os::Union{String, Nothing},
+        version::Union{Vector{VersionNumber}, Nothing},
+        else_version::Union{VersionNumber, Nothing})
```
I have the defaults at the top level in `BotConfg`. Do you think I should have them here too?

aminya commented on src/bot/precompile_include.jl:
```diff
@@ -0,0 +1,247 @@
+
+
+"""
+Helper function for multiversion code generation
```
Yes. Does it need rewording?

aminya commented on src/bot/snoopi_bench.jl:
```diff
@@ -1,143 +1,263 @@
 """
-macro snoopi_bench(packageName::String, snoopScript::Expr)
+macro snoopi_bench(configExpr, snoop_script)
+    @warn "This method isn't recommend. Use `@snoopi_bench(config::BotConfig, path_to_exmple_script::String)` instead."
```
Yes, I did not dispatch directly. I `evaluated` them before doing this. But anyway, I think these macros are deprecated. I am using functions instead.

aminya commented on src/bot/snoopi_bench.jl:
```diff
@@ -1,143 +1,263 @@
+    Don't use this for complex expressions. The functionality isn't guaranteed. Especially if you
+    - interpolate into it
+    - use macros directly inside it
+"""
```
Please see https://github.com/timholy/SnoopCompile.jl/pull/71#issuecomment-625279263

timholy commented:
> If you want, we can deprecate the macros right now in the coming version.

Maybe I'm still not clear (I've not really been a user of this facility yet), but the question is, since this isn't merged do we need the non-recommended macros at all? The only reason to need them is for backwards compatibility, but I don't want to add *new* functionality that is non-recommended. Let's get rid of that before this merges.

aminya commented:
I deprecated them. The files are cleaned. ♻️

timholy commented:
OK, I'll review this in the morning and hopefully we can merge.

aminya commented:
> OK, I'll review this in the morning and hopefully we can merge.

That's awesome. 

The only minor aesthetic thing that I was thinking of was using [Suppressor.jl](https://github.com/JuliaIO/Suppressor.jl) to suppress the excess output in `snoopi_bench`. However, I am not sure if `Suppressor.jl` is reliable enough or has any effect on the performance.

aminya commented:
I added the support for `snoopc`, and Added `snoop_bot/bench`, which automatically chooses the snooping method. There is also `snoopv_bench` which uses `@timev`.

I just need to add `timesum` for `snoopc` data.

timholy commented:
Perhaps I should have advertised this before, but my "review" is actually a pull request against this branch. See https://github.com/aminya/SnoopCompile.jl/pull/1

aminya commented:
@timholy If you agree let's merge this into the master branch. This PR is so big, and managing it is getting out of the hands. 

The only two remaining things are:
- the default value for `tmin` is still `0.0` in both `@snoopi` and `BotConfig`. We discussed it here:
https://github.com/aminya/SnoopCompile.jl/pull/1#discussion_r438829417

- Parsing os and version from the yml file. I can add this in a separate PR.
https://github.com/aminya/SnoopCompile.jl/pull/1#discussion_r438176261

timholy commented:
Tada! :fireworks: Thank you for your patience and persistence!

aminya commented:
Yay!
![](https://media1.giphy.com/media/KEG5UtvXUD7WPIhhuy/giphy.gif)

You're welcome!
